### ISSUE TITLE
[nfs-provisioner] Stale file handles on GKE with regional persistent disks

### ISSUE THREAD

**Issue by thirdeyenick** [kubernetes-incubator/external-storage#1212]
Hi,
we are running the nfs-provisioner on a GKE regional cluster. We are also using a regional persistent disk as a storage backend for the NFS provisioner pod. According to the google documentation [1] this provides a persistent disk replicated in 2 zones. In an emergency case the nfs-provisioner pod can just be started on a node in the replicated zone and all should be fine (as the replicated disk can be mounted on the new node). We are using NFS v4.1 which should help a lot in cases where the state of clients has to be recovered.

When we tested the failover scenario we sometimes noticed some 'NFS: stale file handle' errors on connected NFS clients. The clients are writing constantly to a log file. As this was not happening all the time, we started to investigate more. What we have seen so far is:

* when the new nfs-provisioner pod gets scheduled on the same node as before, the clients can just continue to work (there is only a short interruption)
* when the nfs-provisioner pod gets scheduled onto another node (with access to the replicated persistent disk) it does start normally, but clients are experiencing 'NFS: stale file handle' issues. After we restart the client pods everything works normal again.

As this does not provide a real HA solution, we are searching for explanations/solutions here :-) Does anyone have an idea why this happens?

We are using the gke version 1.12.7-gke.25.

[1] https://cloud.google.com/blog/products/gcp/get-higher-availability-with-regional-persistent-disks-on-google-kubernetes-engine

**Comment by msau42** [comment]
/assign
I'll try to find some answers, but I suspect this is related to how the NFS client handles connections to the server.  I saw similar reports using NFS backed by EmptyDir: https://github.com/kubernetes/kubernetes/issues/75918

Have you tried regular PD to back the NFS share? Does it also see the same issue?

**Comment by thirdeyenick** [comment]
Thx for your answer. That actually inspired me to do some more debugging on the topic :-)

So I tested to restart the nfs provisioner on a different node in the same zone with the same google persistent disk and all clients got into the 'stale file handle' state again. This lead to the assumption that whenever the underlying host changes we run into the issue. I enabled debug logs in ganesha (MID_DEBUG level) and found the following line in the case of stale file handles:
```
26/08/2019 08:32:39 : epoch 5d639925 : nfs-server-provisioner-nfs-r9fhsa-0 : nfs-ganesha-19[svc_12] vfs_extract_fsid :FSAL :M_DBG :Handle len 18 0x45: fsid=0x0000003000000008.0x0000000000000030, type 0x01, opaque: (8:0x010004005f00b991)
26/08/2019 08:32:39 : epoch 5d639925 : nfs-server-provisioner-nfs-r9fhsa-0 : nfs-ganesha-19[svc_12] vfs_check_handle :FSAL :INFO :Could not map fsid=0x0000000000000008.0x0000000000000030 to filesystem
26/08/2019 08:32:39 : epoch 5d639925 : nfs-server-provisioner-nfs-r9fhsa-0 : nfs-ganesha-19[svc_12] complete_op :NFS4 :DEBUG :Status of OP_PUTFH in position 0 = NFS4ERR_STALE, op response size is 4 total response size is 40
```

I read the corresponding part of the NFS4.1 RFC [1] and it looks like that whatever handle the client got from the server needs to be valid for the lifetime of the filesystem object (even after restarts of the NFS server).

After that I found that 'fsid_device' is always set to true in the ganesha configuration [2]. This leads to server generated file handles based on the major/minor device ID of the mounted storage block device (the one which backs '/export'). As these device IDs differ between different gke nodes, the server expects a different file handle when being restarted on a different node and returns NFS4ERR_STALE. As there is already a file system ID generated per export configuration I would be interested in the reason of setting fsid_device to 'true'?

I created a new docker image with 'fsid_device' set to false and also with the most current ganesha version (2.8.2) as 2.7.x will not get any fixes back ported [3]. With that a restart of the nfs provisioner was possible. Clients only experienced a wait time of 90s (the grace period).

Would it be possible to make the 'fsid_device' setting customizable?

[1] https://tools.ietf.org/html/rfc3010#section-4.2.2
[2] https://github.com/kubernetes-incubator/external-storage/blob/350c374b1331b4b636f30f3d41c87d76844dc8d3/nfs/pkg/server/server.go#L114
[3] https://github.com/nfs-ganesha/nfs-ganesha/wiki#Release_Status

**Comment by thirdeyenick** [comment]
We are using the nfs-provisioner in a regional google GKE cluster. As far as I know docker is not used as the container runtime, but I might be wrong (did not find documentation about what is used underneath). Having the possibility to set 'fsid_device' would solve the issue for us. We just need to get the change as well into the upstream helm chart. After that we are fine. In the end the combination of nfs provisioner + google regional persistent disk provides a nice HA solution (we might need to lower the grace period a bit though).

On a different note: Could you tell me something about the update strategy with the Ganesha version in use? If I understand the wiki page of ganesha correctly, then the 2.7.x version will not get fixes and 2.8.x should be used instead. Should this be a separate issue? There is also a minimal version of fedora available which could be used with a multistage build. In my tests I got a 100MB smaller image in the end.

**Comment by thirdeyenick** [comment]
All of this is solved and merged. Will close this issue.
